### PR TITLE
Energy computed instead of reading

### PR DESCRIPTION
**Pull request by jotakar**
Some modules reads energy in kwh with precission of 1 digit  (2,1 kwh, 2,2 kwh ...) but sometimes the energy consumed in a day is less than 0.1 kwh. 
Think about a lamp type led or low consum: 8 watts for 10 hours is 80 wh (0,080kwh). but module reports 0.0 kwh for this day. 
With this modification system gets the energy value by software: it reads the usage and the interval of time and make a simple operation:  usage*time (usage: watts, time: minutes). 
I use field Options for Energy Meter devices. If  enabled (1)  the energy is calculated by program if  disabled (0) energy value is read from the device. 
I've found this issue with Qubino relays.

### CODE REVIEW

**Comment by gizmocuz** [comment]
Thank you for the patch, but as you might have noticed both our build systems have problems with compile errors.

**Comment by gizmocuz** [comment]
You can not change the language files, please provide here (as text) the needed changes to the language file

**Comment by jotakar** [comment]
The words added and their translation into spanish are:
Energy read: Lectura de energía
Computed: Calculada
From device: Desde el módulo

Thanks

**Comment by gizmocuz** [comment]
Why did you remove the 'HTTPPostContentType' in SQLHelper.cpp ?

File: main/SQLHelper.cpp [hunk 1]
```diff
@@ -3095,8 +3090,9 @@ unsigned long long CSQLHelper::UpdateValueInt(const int HardwareID, const char*
 	bool bDeviceUsed = false;
 	bool bSameDeviceStatusValue = false;
 	std::vector<std::vector<std::string> > result;
-	result = safe_query("SELECT ID,Name, Used, SwitchType, nValue, sValue FROM DeviceStatus WHERE (HardwareID=%d AND DeviceID='%q' AND Unit=%d AND Type=%d AND SubType=%d)",HardwareID, ID, unit, devType, subType);
-	if (result.size()==0)
+	result = safe_query("SELECT ID,Name, Used, SwitchType, nValue, sValue, LastUpdate, Description, Options FROM DeviceStatus WHERE (HardwareID=%d AND DeviceID='%q' AND Unit=%d AND Type=%d AND SubType=%d)",HardwareID, ID, unit, devType, subType);
```
Comment (gizmocuz) [inline main/SQLHelper.cpp]: Do we need to query for Description?

File: main/SQLHelper.cpp [hunk 2]
```diff
@@ -3455,7 +3473,7 @@ unsigned long long CSQLHelper::UpdateValueInt(const int HardwareID, const char*
 		CheckSceneStatusWithDevice(ulID);
 		break;
 	}
-	if (bDeviceUsed)
+if (bDeviceUsed)
```
Comment (gizmocuz) [inline main/SQLHelper.cpp]: missing a tab here... (that was previously here)

File: main/SQLHelper.cpp [hunk 3]
```diff
@@ -4017,8 +4035,11 @@ void CSQLHelper::UpdateRainLog()
 			if (splitresults.size()<2)
 				continue; //impossible
 
-			int rate=atoi(splitresults[0].c_str());
-			float total = static_cast<float>(atof(splitresults[1].c_str()));
+			float total=0;
+			int rate=0;
+
+			rate=atoi(splitresults[0].c_str());
+			total = static_cast<float>(atof(splitresults[1].c_str()));
```
Comment (gizmocuz) [inline main/SQLHelper.cpp]: revert these changes please, they are not needed and only give cppcheck issues
This patch is about energy/power anyway ;)

File: main/SQLHelper.cpp [hunk 4]
```diff
@@ -4126,10 +4147,7 @@ void CSQLHelper::UpdateUVLog()
 	GetPreferencesVar("SensorTimeout", SensorTimeOut);
 
 	std::vector<std::vector<std::string> > result;
-	result=safe_query("SELECT ID,Type,SubType,nValue,sValue,LastUpdate FROM DeviceStatus WHERE (Type=%d) OR (Type=%d AND SubType=%d)", 
-		pTypeUV,
-		pTypeGeneral, sTypeUV
-	);
+	result=safe_query("SELECT ID,Type,SubType,nValue,sValue,LastUpdate FROM DeviceStatus WHERE (Type=%d)", pTypeUV);
```
Comment (gizmocuz) [inline main/SQLHelper.cpp]: Revert this ? Why did you do this ?

File: main/SQLHelper.cpp [hunk 5]
```diff
@@ -4169,7 +4187,7 @@ void CSQLHelper::UpdateUVLog()
 			//insert record
 			safe_query(
 				"INSERT INTO UV (DeviceRowID, Level) "
-				"VALUES ('%llu', '%g')",
+				"VALUES ('%llu', '%.1f')",
```
Comment (gizmocuz) [inline main/SQLHelper.cpp]: revert this, why did you do this ?

File: main/SQLHelper.cpp [hunk 6]
```diff
@@ -5388,7 +5406,7 @@ void CSQLHelper::AddCalendarUpdateUV()
 			//insert into calendar table
 			result=safe_query(
 				"INSERT INTO UV_Calendar (DeviceRowID, Level, Date) "
-				"VALUES ('%llu', '%g', '%q')",
+				"VALUES ('%llu', '%.2f', '%q')",
```
Comment (gizmocuz) [inline main/SQLHelper.cpp]: Revert this, why did you do this ?

File: main/SQLHelper.cpp (hunk 1 above)
Comment (jotakar) [inline main/SQLHelper.cpp]: No, I had to have deleted Description. I'll delete it.

File: main/SQLHelper.cpp (hunk 3 above)
Comment (jotakar) [inline main/SQLHelper.cpp]: Sorry, I uploaded my version of SQLHelper.cpp and it had this code, so I reverted some changes made after my versión (5231). I'll revert all code I did'nt modified. You're right: this patch is just for energy/power.
Sorry it's my first pull.

File: main/SQLHelper.cpp (hunk 4 above)
Comment (jotakar) [inline main/SQLHelper.cpp]: Like the upper comment.

File: main/SQLHelper.cpp (hunk 3 above)
Comment (gizmocuz) [inline main/SQLHelper.cpp]: No problem... ;)

File: main/SQLHelper.cpp [hunk 7]
```diff
@@ -5388,7 +5411,7 @@ void CSQLHelper::AddCalendarUpdateUV()
 			//insert into calendar table
 			result=safe_query(
 				"INSERT INTO UV_Calendar (DeviceRowID, Level, Date) "
-				"VALUES ('%llu', '%g', '%q')",
+				"VALUES ('%llu', '%.g', '%q')",
```
Comment (gizmocuz) [inline main/SQLHelper.cpp]: I still see a lot of non-needed changes, and for example this one (procent point 'g') is not what it should be
Maybe best to revert this whole file, and only make that change needed for your patch
So we also do not get extra lines, or extra spaces at the end of the lines

File: main/SQLHelper.cpp [hunk 8]
```diff
@@ -3132,17 +3133,40 @@ unsigned long long CSQLHelper::UpdateValueInt(const int HardwareID, const char*
...
+	            ntime.tm_min=atoi(sLastUpdate.substr(14,2).c_str());
+	            ntime.tm_sec=atoi(sLastUpdate.substr(17,2).c_str());
+	            interval = now - mktime(&ntime);
+		    StringSplit(result[0][5].c_str(), ";", parts);            
+	            nEnergy = strtof(parts[0].c_str(),NULL)*interval/3600 + strtof(parts[1].c_str(), NULL);
+	            
+	            StringSplit(sValue, ";", parts);
+	            sprintf(sCompValue, "%s;%.0f", parts[0].c_str(),nEnergy);
+	            sValue = sCompValue;
+            }
```
Comment (gizmocuz) [inline main/SQLHelper.cpp]: How do you know it is an Energy device here ? All devices can have sOption="1"

**Comment by jotakar** [comment]
I can't see why Travis C says there is an error. I see 
`Error: Failed to download resource "boost"`
But I can't see anything about my file ... What must I do?

File: main/SQLHelper.cpp (hunk 8 above)
Comment (jotakar) [inline main/SQLHelper.cpp]: Well I can use devType and subType , that's what I've done in last modif.

**Comment by jotakar** [comment]
Hi
Which is the status of this pull request?

**Comment by jotakar** [comment]
Hi
I'm using this modification in my net and it's working fine. 
Is it being tested by anyone else?

**Comment by YcKe** [comment]
Seems like this would solve indeed the Qubino metering precision issue. Great work jotakar! Hope it gets merged soon.

**Comment by jotakar** [comment]
Thanks
I suppose it's being tested or so. I'm using it and it works fine in my net.

**Comment by gizmocuz** [comment]
Thanks

**Comment by jotakar** [comment]
I am pleased to participate in this program.
